### DISCUSSION TITLE
New interface for the SDPA solver

### DISCUSSION THREAD

**Discussion by GauthierGar** [cvxpy/cvxpy#1419] [Q&A] [answered]
Hi,
I would like to discuss the possibility of developing a new interface for the SDPA solver.
There is a lack of free SDP solver that don't rely on matlab and that are available under a permissive
license.

I think that SDPA could be a solution:

1. The users install sdpa from the sourceforge repository (binaries for windows users, source for Mac and Linux users)
2. Then define the SDP problem with cvxpy. The problem has to be process in the sdpa format (standard form)
3. To solve the license issue, SDPA should be started through a system call `subprocess.run()`

This will allow for very convenient implementation of signal processing / super-resolution algorithms
that rely on SDP optimization, even outside the academic field.

I am likely to tackle this issue if after discussion this seems realisable.
Thank you!

**Comment by SteveDiamond** [comment]
Could you say more about the license issues?

**Reply by rileyjmurray** [reply]
@SteveDiamond I believe @GauthierGar is referring to the fact that SDPA is licensed under the GPL. Some interpretations of the GPL require that non-GPL software access it only through certain limited means. We've interpreted the GPL as allowing access through any API, so long as CVXPY doesn't make the GPL software an installation requirement. (We also take liberties with ECOS, due to its historical connection to CVXPY.)

**Reply by GauthierGar** [reply]
Yes, sorry I wasn't clear
Nearly all sdp solvers are licensed under GPL so it's complex to implement some algorithms inside a proprietary software, especially without matlab or proprietary solvers.
This is why I suggest this odd solution with SDPA. SCS is available under MIT but it's not adequate when high numerical precision is required.

I think that it goes against the license's terms to link (or import) a GPL library inside an Apache package:
https://www.gnu.org/licenses/gpl-faq.en.html#MereAggregation
https://opensource.stackexchange.com/questions/6033/can-a-non-gpl-python-program-use-gpl-python-module

A solution can be of asking for the owner to release their solver under a specific license for a given project.

**Reply by usamamuneeb** [reply]
Regarding the licensing issues, the main issue I see is not calling via command line vs importing through the `import` statement. It's more so whether the "semantics of the communication are intimate enough, exchanging complex internal data structures", as mentioned in the first link that @GauthierGar posted.

In my understanding, communication being intimate enough would mean that the calling software is designed to intricately match the called software which I don't believe is the case (the data being exchanged is pretty much formatted the same standard way as also used by SeDuMi and SDPT3). From another perspective, if the communication was intimate enough, the calling software would be reliant on the called software (while `SDPA` in this case would be an optional install, as @rileyjmurray mentioned).

They mention that pipes, sockets and **command-line arguments** are communication mechanisms normally used between two separate programs. `sdpa` can very well be interfaced directly with the binary through external system call rather than it's Python wrapper, if we choose to avoid the `import` technique.

Regarding the point on "running in a shared address space", it's also optional. `SDPA` can read data exported by CVXPY from a file.

**Reply by usamamuneeb** [reply]
Besides licensing, another issue with [SDPA](http://sdpa.sourceforge.net/) was the fact that it’s official Python wrapper (SDPAP) is written for Python 2 (and hasn’t been updated for a while). Part of this wrapper code is C++ (and uses Python 2 C++ headers).

Since CVXPY is in Python 3, I have created this Python 3 port of SDPAP: https://github.com/usamamuneeb/sdpa-python

If there aren't any issues with licensing, I think I can take care of writing the CVXPY interface.

**Reply by SteveDiamond** [reply]
I am not concerned about licensing issues in this case. @usamamuneeb if you're interested in writing the CVXPY interface, that would be very much appreciated! Check out our developer docs on solver interfaces for guidance: https://www.cvxpy.org/contributing/index.html#solver-interfaces

**Reply by usamamuneeb** [reply]
That's great. I will create a PR when I have a preliminary version of the interface ready. Should I use `import` the Python wrapper I linked above or should I externally call the SDPA binary?

The wrapper comes with many technical benefits, including ease of installation for the user.

**Reply by SteveDiamond** [reply]
You can import the python package.

**Reply by usamamuneeb** [reply]
Perfect!

**Comment by usamamuneeb** [comment, marked as answer]
I have finally created a [pull request](https://github.com/cvxpy/cvxpy/pull/1769). Besides that, I had a few questions (even though they aren’t all CVXPY development tasks per se).

1. How do we expect the user to install SDPA? Should it remain [from source](https://sdpa-python.github.io/docs/installation/) or should we create `pip` and `conda` packages linked against all common BLAS/LAPACK implementations? Do we need original author’s permission before submitting to `pip` and `conda`?
2. While the SDPA manuals mention SOCP support, SOCP constraints aren’t really supported by SDPA. In the SDPA (not SDPA-P) code, the relevant functions raise an error `current version does not support SOCP`. Should we convert these constraints to PSD in the CVXPY interface or leave out SOCP support altogether? (SDPA is not really an SQLP solver, it’s an SDP solver).

Lastly, there are some [unimplemented routines](https://sdpa-python.github.io/docs/contributing/) in the SDPA-P package (of which I made the [Python 3 port](https://github.com/sdpa-python/sdpa-python) to be called by CVXPY). These routines are optional but implementing them seems feasible in a reasonable amount of time. They are meant to allow the user to optionally exploit sparsity in the problem and to use an alternate internal conversion method (SeDuMi Primal instead of SeDuMi Dual). I have added comments in the submitted interface that explain the differences between conversion methods.

**Reply by SteveDiamond** [reply]
Thanks for the PR! If you could create pip and conda packages that would be ideal. What is the license for SDPA? I couldn't find it on their website. If it's open source you have the right to make these distributions IMO. Regardless, probably best to email the authors first though.

I would suggest implementing the interface without SOCP constraints first. You can add them in as a second PR if desired.

The unimplemented routines sound valuable but lets not let that be a barrier to getting an initial interface up. It's fine to do multiple passes on the interface.

**Reply by usamamuneeb** [reply]
Certainly, I'll contact the authors regarding the `pip` and `conda` issue. SDPA is GPL v2.0 licensed. And I agree we should start off with the functionality that is natively available for the solver. Unimplemented routines are improvements and not an issue for now.

**Reply by usamamuneeb** [reply]
It took a little more parameter tuning and LP5 accuracy issue is also resolved. All unit tests are now passing.

**Reply by usamamuneeb** [reply]
The authors have responded positively regarding `pip` and `conda` packages.
